ext/storage_sources/s3_store/s3_connection.cpp

```diff
@@ -177,6 +186,35 @@ S3Connection::ObjectExists(const std::string &objectKey, bool &exists, size_t &o
      * Fix later, return a proper error code. Not sure if we always have
      * outcome.GetError().GetResponseCode()
      */
-    std::cerr << "Error in ObjectExists." << std::endl;
     return (1);
 }
+
+/*
+ * BucketExists --
+ *     Checks whether the bucket configured for the class is accessible to us or not.
+ */
+int
+S3Connection::BucketExists(bool &exists) const
+{
+    exists = false;
+
+    Aws::S3Crt::Model::HeadBucketRequest request;
+    request.WithBucket(_bucketName);
+    Aws::S3Crt::Model::HeadBucketOutcome outcome = _s3CrtClient.HeadBucket(request);
+
+    /*
+     * If an object with the given key does not exist the HEAD request will return a 404.
+     * https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadBucket.html Do not fail in this case.
+     */
+    if (outcome.IsSuccess()) {
+        exists = true;
+        return (0);
+    } else if (outcome.GetError().GetResponseCode() == Aws::Http::HttpResponseCode::NOT_FOUND)
+        return (0);
```

Comment:
Is it ok if we get `NOT_FOUND`?

Reply:
Yes, it is expected if the bucket doesn't not exist. Since the function checks for a bucket's existence, it is valid that the bucket doesn't exist. We return that as a boolean, and let the caller decide what they want to do.